lib/api/model/model.dart

```diff
@@ -531,10 +533,68 @@ String? tryParseEmojiCodeToUnicode(String emojiCode) {
   }
 }
 
+/// As in [MessageBase.recipient].
+///
+/// Different from [MessageDestination], this information comes from
+/// [getMessages] or [getEvents], identifying the conversation that contains a
+/// message.
+sealed class Recipient {}
+
+/// The recipient of a stream message.
+@JsonSerializable(fieldRename: FieldRename.snake, createToJson: false)
+class StreamRecipient extends Recipient {
+  StreamRecipient(this.streamId, this.topic);
+
+  int streamId;
+
+  @JsonKey(name: 'subject')
+  TopicName topic;
+
+  factory StreamRecipient.fromJson(Map<String, dynamic> json) =>
+    _$StreamRecipientFromJson(json);
+}
+
+/// The recipient of a DM message.
+class DmRecipient extends Recipient {
+  DmRecipient({required this.allRecipientIds})
+    : assert(isSortedWithoutDuplicates(allRecipientIds.toList()));
+
+  /// The user IDs of all users in the thread, sorted numerically.
+  ///
+  /// This lists the sender as well as all (other) recipients, and it
+  /// lists each user just once.  In particular the self-user is always
+  /// included.
+  ///
+  /// This is required to have an efficient `length`.
+  final List<int> allRecipientIds;
+}
+
+/// A message or message-like object, for showing in a message list.
+///
+/// Other than [Message], we use this for "outbox messages",
+/// representing outstanding [sendMessage] requests.
+abstract class MessageBase<T extends Recipient> {
```

Comment:
api [nfc]: Pull out MessageBase and Recipient

This can be split into two commits:
* One introduces Recipient (or Conversation) and subclasses, and switches StreamMessage and DmMessage to use them;
* One introduces MessageBase.

I think those are fairly conceptually distinct changes, even though the first is a prerequisite for what we want from the second.